### PR TITLE
fix: `EXISTS`/`IN` subqueries hit the count bug for groupless aggregates

### LINKED ISSUE #24960: Correlated `EXISTS`/`IN` subqueries with groupless aggregates hit the count bug
### Describe the bug

A correlated `EXISTS` or `IN` subquery containing a groupless aggregate (e.g. `count(*)`, with no `GROUP BY`) returns incorrect results when the outer row has 0 matches on the subquery.

A groupless aggregate always produces exactly one output row, even over zero input rows. The current decorrelation instead drops the outer row entirely, as if the subquery had produced no rows at all.

### To Reproduce

```sql
CREATE TABLE t1(t1_int int); INSERT INTO t1 VALUES (1),(2),(3);
CREATE TABLE t2(t2_int int); INSERT INTO t2 VALUES (2);

SELECT t1.t1_int FROM t1 WHERE EXISTS (
    SELECT count(*) FROM t2 WHERE t1.t1_int = t2.t2_int
) ORDER BY 1;
```

### Expected behavior

The query above should return all three rows: `1, 2, 3`.

For `t1_int=1` and `t1_int=3`, no row in `t2` matches, so the inner `count(*)` evaluates to `0` rather than producing no row at all. A `0` is still a row, so `EXISTS` should be `true` for those two outer rows too.

Currently, DataFusion returns a single row, `2`, silently dropping the two rows whose `count(*)` evaluated to `0`.

The `NOT EXISTS` variant of the same query surfaces the same underlying bug. It should return no rows (since `EXISTS` is true for every row) but instead returns `1, 3`.

### Additional context

Same bug class referenced in #10553 and #15032/#15281.

### PR DESCRIPTION
**Pull request by mohammadnaqvi04**
## Which issue does this PR close?

- Closes #24960.

## Rationale for this change

See #24960.

## What changes are included in this PR?

Correlated `EXISTS`/`NOT EXISTS`/`IN`/`NOT IN` subqueries had count-bug compensation available but never turned it on.

* Turn on count-bug compensation for correlated `EXISTS`/`NOT EXISTS`/`IN`/`NOT IN` subqueries.
* When compensation is needed, decorrelate into a `LEFT JOIN` and substitute the correct value instead of dropping the row.
* Fix a related bug in the `IN` case, where the comparison value could wrongly affect which rows the join matched, corrupting results for rows with a real correlated match.
* Make unsupported join shapes fail loudly instead of silently falling back to incorrect behavior.

## What is the testing strategy for this PR?

Added eighteen sqllogictest cases in `subquery.slt` and one unit test.

They cover `EXISTS`, `NOT EXISTS`, `IN`, and `NOT IN`. Each is tested standalone and embedded in an `OR`. Each is also tested against an aggregate with a zero default and one with a `NULL` default.

Ran the full `datafusion`, `datafusion-cli`, and `datafusion-sqllogictest` suites, and `./dev/rust_lint.sh`.

## Are there any user-facing changes?

None, except in the sense that this fixes a user-visible bug.

## AI Usage

_AI was used during development of this PR, primarily to help review the test suite and verify test cases against the implemented fix._

### CODE REVIEW

File: datafusion/sqllogictest/test_files/subquery.slt [hunk 1]
```diff
@@ -1138,6 +1138,83 @@ select t1.t1_int from t1 where (
...
+SELECT t1_id, t1_int FROM t1 WHERE 0 NOT IN (SELECT count(*) FROM t2 WHERE t1.t1_id = t2.t2_id)
+----
+11 1
+22 2
+44 4
+
+# correlated_not_in_groupless_count_agg_count_bug_having
+query II
+SELECT t1_id, t1_int FROM t1 WHERE 0 NOT IN (SELECT count(*) FROM t2 WHERE t1.t1_id = t2.t2_id HAVING count(*) = 0)
+----
```
Comment (AdamGS) [inline datafusion/sqllogictest/test_files/subquery.slt:1284]: This is still buggy, it should return:
```
11 1
22 2
44 4
```

File: datafusion/sqllogictest/test_files/subquery.slt [hunk 2]
```diff
@@ -1370,6 +1447,71 @@ where t1_id > 40 or not exists (select 1 from t2 where t2.t2_int > t1.t1_int)
...
+where t1_id > 40 or 0 in (select sum(t2_int) from t2 where t1.t1_id = t2.t2_id)
+----
+44
+
+# correlated_not_in_groupless_count_agg_count_bug_embedded_in_or
+query I rowsort
+select t1_id from t1
+where t1_id > 40 or 0 not in (select count(*) from t2 where t1.t1_id = t2.t2_id having count(*) = 0)
+----
+44
```
Comment (AdamGS) [inline datafusion/sqllogictest/test_files/subquery.slt:1582]: This result is also wrong, it should return:
```
11 1
22 2
44 4
```

**Review by AdamGS — COMMENTED** [review]
I think there are still bugs here, see comments. I've compared the results against DuckDB.

File: datafusion/sqllogictest/test_files/subquery.slt (hunk 1 above)
Comment (mohammadnaqvi04) [reply datafusion/sqllogictest/test_files/subquery.slt:1284]: Oof, two things here:

1. The join mishandled `NULL` negation for `NOT IN`
2. This test's expected value was wrong too, they just happened to match each other. 

Fixed both in the follow-up commit.

File: datafusion/sqllogictest/test_files/subquery.slt (hunk 2 above)
Comment (mohammadnaqvi04) [reply datafusion/sqllogictest/test_files/subquery.slt:1582]: Fixed both in followup.

```
> SELECT t1_id, t1_int FROM t1 WHERE 0 NOT IN (SELECT count(*) FROM t2 WHERE t1.t1_id = t2.t2_id HAVING count(*) = 0);

+-------+--------+
| t1_id | t1_int |
+-------+--------+
| 11    | 1      |
| 22    | 2      |
| 44    | 4      |
+-------+--------+
3 row(s) fetched.
Elapsed 0.018 seconds.

> select t1_id from t1 where t1_id > 40 or 0 not in (select count(*) from t2 where t1.t1_id = t2.t2_id having count(*) = 0);

+-------+
| t1_id |
+-------+
| 11    |
| 22    |
| 44    |
+-------+
3 row(s) fetched.
Elapsed 0.011 seconds.
```

**Comment by mohammadnaqvi04** [comment]
@AdamGS Thanks for the review! It seems like there was a trickier bug hiding under the one I initially wanted to fix. `HAVING`-failed rows were compensated with `NULL`, and negating `NULL` for `NOT IN` caused rows to be dropped rather than included. Fixed now.

**Comment by AdamGS** [comment]
I did a pass with claude, and it spit out the following tests that all pass on `main` and in DuckDB, but fail on this branch:
```sql
statement ok
CREATE TABLE outer_rows(k INT);

statement ok
INSERT INTO outer_rows VALUES (1);

statement ok
CREATE TABLE inner_rows(k INT, v INT);

statement ok
INSERT INTO inner_rows VALUES (1, NULL);

statement ok
CREATE TABLE text_keys(k VARCHAR);

statement ok
INSERT INTO text_keys VALUES ('1'), ('01');

# HAVING evaluates to NULL for the matching group and must discard its row.
query I
SELECT o.k
FROM outer_rows o
WHERE EXISTS (
  SELECT count(*) FROM inner_rows i WHERE i.k = o.k
  HAVING count(*) = 0 OR sum(i.v) > 0
);
----

# The same NULL-valued HAVING makes the subquery empty, so NOT IN is true.
query I
SELECT o.k
FROM outer_rows o
WHERE 1 NOT IN (
  SELECT count(*) FROM inner_rows i WHERE i.k = o.k
  HAVING count(*) = 0 OR sum(i.v) > 0
);
----
1

# LIMIT 0 removes the aggregate row, so EXISTS is false.
query I
SELECT o.k
FROM outer_rows o
WHERE EXISTS (
  SELECT count(*) FROM inner_rows i WHERE i.k = o.k LIMIT 0
);
----

# Both text keys cast to 1, but EXISTS must return the outer row only once.
query I
SELECT o.k
FROM outer_rows o
WHERE EXISTS (
  SELECT count(*) FROM text_keys i WHERE CAST(i.k AS INT) = o.k
);
----
1
```
I didn't have the time to try and dive into the individual causes here, but might be worth it to try and split this PR into smaller ones that are more specific (and potentially stacked) as each one seems to be pretty hard to review

**Comment by mohammadnaqvi04** [comment]
@AdamGS Thanks for the second pass. I worked through the faulty queries and am re-scoping what this PR is and isn't claiming to fix.

```sql
# HAVING evaluates to NULL for the matching group and must discard its row.
query I
SELECT o.k
FROM outer_rows o
WHERE EXISTS (
  SELECT count(*) FROM inner_rows i WHERE i.k = o.k
  HAVING count(*) = 0 OR sum(i.v) > 0
);
----

# The same NULL-valued HAVING makes the subquery empty, so NOT IN is true.
query I
SELECT o.k
FROM outer_rows o
WHERE 1 NOT IN (
  SELECT count(*) FROM inner_rows i WHERE i.k = o.k
  HAVING count(*) = 0 OR sum(i.v) > 0
);
----
1
```

These two cover the same three-valued bug where a HAVING predicate can itself be `NULL`. Both are now fixed.

The `LIMIT 0` case and the CAST-based duplicate-match case are gaps in what count-bug compensation is capable of tracking. LIMIT-truncated rows aren't tracked the way HAVING-failed rows are, and the compensation join can fan out multiple rows per outer row under a non-injective (one-to-many from outer row to inner) correlation key. I gave this some thought, and think both are big enough that they don't belong in this PR.

I've decided to narrow this PR's scope to `EXISTS`/`NOT EXISTS`/`IN`/`NOT IN` over groupless aggregates with `HAVING`, including `NULL`-valued and compound `HAVING` predicates. I'll open LIMIT-truncation and join fan-out as separate follow-ups.

Also, in the process of more adversarial review targeting the `HAVING` mechanics here, I discovered three more. I've included fixes and test cases for all three:

1. A second filter stacked on an already-compensated aggregate fell back to silently dropping the HAVING clause.
2. The unmatched-row default for that stacked-filter case skipped the combined predicate on the `IN`/`NOT IN` side.
3. The unmatched-row default for that stacked-filter case skipped the combined predicate on the `EXISTS` side.

I've also just now run and compared every new query I added in `subquery.slt` from my branch against latest DuckDB and DataFusion `main` builds.

**Comment by AdamGS** [comment]
Seems like you ran into some conflicts.

But as you mentioned, seems like the last two queries from above still show a regression, I'm not sure how I feel about knowingly introducing correctness regressions. I'll try and do a deeper pass over the PR sometime this week as I think about that, maybe someone else will also weigh in.

**Comment by mohammadnaqvi04** [comment]
I took a look at potentially splitting this PR into two more contained ones between `EXISTS`/`NOT EXISTS`, and then a stacked `IN`/`NOT IN` on top, to reduce some of the surface area here.

The reason I'm holding off on doing so, for now at least, is because the non-injective correlation bug, from your last failing example query, appears in both of the aforementioned shape categories. I'm going to stress test a bit more, and I can refactor for better composability afterward.

On a secondary note, to make this a bit more robust, I'm working on extending our [fuzz test setup](https://github.com/apache/datafusion/tree/main/datafusion/core/tests/fuzz_cases) to generate random query shapes for groupless aggregates. I'll cross-verify against other engines to see where this breaks, and hopefully iterate from there.

**Comment by mohammadnaqvi04** [comment]
@AdamGS Circling back on the `LIMIT 0` and fan-out cases from your last comment, both are fixed now. I changed how the subquery's aggregate is grouped so that two raw values that equal eachother post-cast don't fan-out anymore. 

```sql
> SELECT o.k
FROM outer_rows o
WHERE EXISTS (
  SELECT count(*) FROM inner_rows i WHERE i.k = o.k LIMIT 0
);
+---+
| k |
+---+
+---+
0 row(s) fetched.
Elapsed 0.005 seconds.

> SELECT o.k
FROM outer_rows o
WHERE EXISTS (
  SELECT count(*) FROM text_keys i WHERE CAST(i.k AS INT) = o.k
);
+---+
| k |
+---+
| 1 |
+---+
1 row(s) fetched.
Elapsed 0.006 seconds.
```
I'm going to split this into two smaller PRs for reviewability: one for `EXISTS`/`NOT EXISTS`, then `IN`/`NOT IN` afterward once that one's merged. I'll close this PR once the first one is up.

**Comment by mohammadnaqvi04** [comment]
Closing in favor of https://github.com/apache/datafusion/pull/25391.
